feat: fix stderrthreshold not honored when logtostderr is set (#212) + two new flags

pierluigilenoci opened the pull request:
## Description

This PR fixes a long-standing issue (#212) where the `-stderrthreshold` flag is completely ignored when `-logtostderr=true`, preventing users from filtering logs by severity level when logging to stderr.

The solution introduces two new flags while maintaining **100% backward compatibility** through opt-in behavior.

## What type of PR is this?

/kind feature
/kind bug

## What this PR does / why we need it

### Problem
When `-logtostderr=true` is set, klog writes all logs to stderr regardless of the `-stderrthreshold` setting. This causes excessive noise in log outputs and makes it impossible to filter logs by severity, which is particularly problematic for applications that want to reduce log verbosity (see #212 and https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/387).

### Solution
This PR adds two new flags:

1. **`-legacy_stderr_threshold_behavior`** (default: `true`)
   - `true` (default): Maintains current behavior - all logs go to stderr when `-logtostderr=true`
   - `false`: New behavior - honors `-stderrthreshold` even when `-logtostderr=true`

2. **`-alsologtostderrthreshold`** (default: `INFO`)
   - Provides severity filtering specifically for `-alsologtostderr=true`
   - Default value maintains backward compatibility (all logs mirror to stderr)
   - Can be set to WARNING or ERROR for finer control

### Additional Fix
Fixed a bug in `severityValue.Set()` where it was hardcoded to always set `logging.stderrThreshold`, preventing multiple `severityValue` instances from working independently.

## Usage Examples

**Enable severity filtering (new behavior):**
```bash
./myapp -logtostderr=true \
        -legacy_stderr_threshold_behavior=false \
        -stderrthreshold=ERROR
```
Result: Only ERROR and FATAL messages appear in stderr.

**Filter logs when mirroring to stderr:**
```bash
./myapp -logtostderr=false \
        -alsologtostderr=true \
        -alsologtostderrthreshold=WARNING \
        -log_dir=/var/log/myapp
```
Result: All logs in files, only WARNING+ to stderr.

## Which issue(s) this PR fixes

Fixes #212

## Special notes for your reviewer

- **Backward compatibility**: All existing behavior is preserved by default. The new functionality is strictly opt-in.
- **Testing**: Added comprehensive unit tests covering both legacy and new behavior
- **No breaking changes**: All existing applications continue to work without modification

## Does this PR introduce a user-facing change?

```release-note
Add optional severity filtering for stderr output when -logtostderr=true through new -legacy_stderr_threshold_behavior flag (default: true for backward compatibility). Add -alsologtostderrthreshold flag for finer control when using -alsologtostderr=true.
```

## Additional documentation

See the example in `examples/stderr_threshold_fix/main.go` for a practical demonstration.

### Migration Guide

**For end users:** No action required. Default behavior unchanged.

**To opt into severity filtering:**
1. Set `-legacy_stderr_threshold_behavior=false`
2. Use `-stderrthreshold=ERROR` (or WARNING) to filter logs

## Related PRs/Issues

- Original PR (reverted due to breaking changes): #31
- Revert PR: #50  
- Previous attempt: #221
- Related issue: https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/387

pierluigilenoci committed:
Fix stderrthreshold not honored when logtostderr is set

Add two new flags to allow severity filtering with logtostderr while
maintaining full backward compatibility:

- legacy_stderr_threshold_behavior (default: true): Controls whether
  stderrthreshold is honored when logtostderr=true
- alsologtostderrthreshold (default: INFO): Provides filtering control
  when alsologtostderr=true

This fix allows users to opt-in to severity filtering when logging to
stderr, addressing a long-standing issue where all logs would be written
to stderr regardless of severity level.

The default behavior remains unchanged for backward compatibility.

Ref: #212

pierluigilenoci commented:
@harshanarayana @thockin @dims @pohly @mengjiao-liu I'm tagging you to get your feedback.

@aramase @neolit123 @yagonobre @yuzhiquan @serathius @mtaufen @DirectXMan12 @vincepri @shsjshentao @dims @justinsb I'm tagging you too because you were involved in the discussion and attempted to resolve this issue several years ago.

pohly reviewed (commented):
I'm undecided whether it's worth fixing this. But at least now it's not a breaking change, so I am okayish with trying. I haven't looked at all code changes yet.

pohly commented:
@pierluigilenoci: it took you more than two years to come up with this PR since we [last discussed the issue](https://github.com/kubernetes/klog/issues/212#issuecomment-1822442340). It's not urgent, is it? So please understand that reviewers are not immediately jumping on reviewing this. For example, I am on vacation.

pohly commented:
I can't promise anything at this point. IMHO it has low priority and there's always something else that needs my attention more urgently.

pohly reviewed (changes requested):
Some minor nits. I don't see any problem with the logic, but there's always a risk that I am missing something. Therefore I'm still unsure whether it really makes sense to extend klog. We've long said that it's feature set is frozen.

@pierluigilenoci: are you really going to use the new flags or did you just submit the PR to get closure on a years old issue?

@aramase, @deyanp, @tallclair, @kcone, @nyejon: you all commented on https://github.com/kubernetes/klog/issues/212 at some point. Does it really still make sense to fix that issue?

pierluigilenoci committed:
Update klog.go

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

pierluigilenoci committed:
Update stderr_threshold_test.go

Co-authored-by: Patrick Ohly <patrick.ohly@intel.com>

pierluigilenoci committed:
refactor: address code review feedback from @pohly

- Move alsologtostderrthreshold initialization next to stderrThreshold
  at the beginning of init() for better organization
- Simplify conditional logic in output() by combining conditions with
  || operators to avoid redundant .get() calls and improve readability

These changes don't affect functionality, only improve code organization
and efficiency as suggested in the PR review.

pohly reviewed (commented):
/lgtm

Before we merge this I'd like to hear from *someone* else besides @pierluigilenoci  that this really still matters.

@pierluigilenoci: as your next step seems to be submitting changes to other reports, perhaps you can ask the owners of those repos whether they will take your changes?

pierluigilenoci commented:
## Downstream PRs propagating the fix

I've been propagating this fix to downstream projects that use klog v2. Here is the current status (86 PRs total):

### Merged (11)

| Repository | PR |
|-----------|---:|
| kubernetes-sigs/aws-efs-csi-driver | [#1822](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/1822) |
| kubernetes-sigs/azurefile-csi-driver | [#3037](https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/3037) |
| operator-framework/operator-lifecycle-manager | [#3801](https://github.com/operator-framework/operator-lifecycle-manager/pull/3801) |
| kubernetes-sigs/gcp-compute-persistent-disk-csi-driver | [#2291](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/2291) |
| kubernetes-csi/csi-driver-nfs | [#1078](https://github.com/kubernetes-csi/csi-driver-nfs/pull/1078) |
| kubernetes/kube-state-metrics | [#2906](https://github.com/kubernetes/kube-state-metrics/pull/2906) |
| kubernetes-sigs/blob-csi-driver | [#2423](https://github.com/kubernetes-sigs/blob-csi-driver/pull/2423) |
| kubernetes-sigs/gcp-filestore-csi-driver | [#1279](https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/pull/1279) |
| kubernetes/node-problem-detector | [#1267](https://github.com/kubernetes/node-problem-detector/pull/1267) |
| nats-io/nack | [#360](https://github.com/nats-io/nack/pull/360) |
| eryajf/kite-desktop | [#66](https://github.com/eryajf/kite-desktop/pull/66) |

### Closed (8)

| Repository | PR | Reason |
|-----------|---:|--------|
| vapor-ware/k8s-elector | [#30](https://github.com/vapor-ware/k8s-elector/pull/30) | dormant |
| kubernetes-sigs/dranet | [#170](https://github.com/kubernetes-sigs/dranet/pull/170) | closed by maintainer |
| awslabs/mountpoint-s3-csi-driver | [#760](https://github.com/awslabs/mountpoint-s3-csi-driver/pull/760) | closed by maintainer |
| rancher/rke2 | [#10281](https://github.com/rancher/rke2/pull/10281) | closed by maintainer |
| corneliusweig/rakkess | [#181](https://github.com/corneliusweig/rakkess/pull/181) | closed by maintainer |
| superedge/superedge | [#531](https://github.com/superedge/superedge/pull/531) | closed by maintainer |
| openshift/source-to-image | [#1372](https://github.com/openshift/source-to-image/pull/1372) | closed by maintainer |
| cilium/tetragon | [#4903](https://github.com/cilium/tetragon/pull/4903) | duplicate |

### Open (67)

| Repository | PR |
|-----------|---:|
| kubernetes-sigs/secrets-store-csi-driver | [#2016](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/2016) |
| kedacore/keda | [#7568](https://github.com/kedacore/keda/pull/7568) |
| aws/secrets-store-csi-driver-provider-aws | [#589](https://github.com/aws/secrets-store-csi-driver-provider-aws/pull/589) |
| kubernetes/autoscaler | [#9408](https://github.com/kubernetes/autoscaler/pull/9408) |
| kubernetes-sigs/metrics-server | [#1782](https://github.com/kubernetes-sigs/metrics-server/pull/1782) |
| kubernetes-sigs/nfs-subdir-external-provisioner | [#386](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/pull/386) |
| kyverno/kyverno | [#15711](https://github.com/kyverno/kyverno/pull/15711) |
| kubernetes-sigs/kindnet | [#85](https://github.com/kubernetes-sigs/kindnet/pull/85) |
| kubernetes-sigs/sig-storage-local-static-provisioner | [#553](https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/pull/553) |
| kubernetes-sigs/azurelustre-csi-driver | [#278](https://github.com/kubernetes-sigs/azurelustre-csi-driver/pull/278) |
| kubernetes/kube-openapi | [#588](https://github.com/kubernetes/kube-openapi/pull/588) |
| kubernetes/minikube | [#22785](https://github.com/kubernetes/minikube/pull/22785) |
| fluxcd/flagger | [#1910](https://github.com/fluxcd/flagger/pull/1910) |
| cert-manager/cert-manager | [#8738](https://github.com/cert-manager/cert-manager/pull/8738) |
| FairwindsOps/pluto | [#596](https://github.com/FairwindsOps/pluto/pull/596) |
| kubernetes-csi/external-snapshot-metadata | [#230](https://github.com/kubernetes-csi/external-snapshot-metadata/pull/230) |
| volcano-sh/volcano | [#5134](https://github.com/volcano-sh/volcano/pull/5134) |
| kubernetes-sigs/kind | [#4124](https://github.com/kubernetes-sigs/kind/pull/4124) |
| childe/gohangout | [#280](https://github.com/childe/gohangout/pull/280) |
| rancher/fleet | [#4899](https://github.com/rancher/fleet/pull/4899) |
| kcp-dev/kcp | [#3944](https://github.com/kcp-dev/kcp/pull/3944) |
| velero-io/velero | [#9654](https://github.com/velero-io/velero/pull/9654) |
| google/cadvisor | [#3859](https://github.com/google/cadvisor/pull/3859) |
| kubernetes/test-infra | [#36730](https://github.com/kubernetes/test-infra/pull/36730) |
| lyft/flinkk8soperator | [#302](https://github.com/lyft/flinkk8soperator/pull/302) |
| vmware-tanzu/vm-operator | [#1539](https://github.com/vmware-tanzu/vm-operator/pull/1539) |
| int128/kubelogin | [#1556](https://github.com/int128/kubelogin/pull/1556) |
| kubernetes-sigs/node-feature-discovery | [#2494](https://github.com/kubernetes-sigs/node-feature-discovery/pull/2494) |
| google/certificate-transparency-go | [#1785](https://github.com/google/certificate-transparency-go/pull/1785) |
| google/trillian | [#3878](https://github.com/google/trillian/pull/3878) |
| kubernetes/cloud-provider-gcp | [#1080](https://github.com/kubernetes/cloud-provider-gcp/pull/1080) |
| kubernetes/perf-tests | [#3977](https://github.com/kubernetes/perf-tests/pull/3977) |
| koordinator-sh/koordinator | [#2874](https://github.com/koordinator-sh/koordinator/pull/2874) |
| pingcap/tidb-operator | [#6847](https://github.com/pingcap/tidb-operator/pull/6847) |
| apecloud/kubeblocks | [#10169](https://github.com/apecloud/kubeblocks/pull/10169) |
| werf/kubedog | [#391](https://github.com/werf/kubedog/pull/391) |
| kubewharf/katalyst-core | [#1134](https://github.com/kubewharf/katalyst-core/pull/1134) |
| devspace-sh/devspace | [#3221](https://github.com/devspace-sh/devspace/pull/3221) |
| gianlucam76/k8s-cleaner | [#590](https://github.com/gianlucam76/k8s-cleaner/pull/590) |
| mongodb/mongodb-kubernetes | [#1013](https://github.com/mongodb/mongodb-kubernetes/pull/1013) |
| ahmetb/kubectl-pods_on | [#47](https://github.com/ahmetb/kubectl-pods_on/pull/47) |
| tinkerbell/cluster-api-provider-tinkerbell | [#561](https://github.com/tinkerbell/cluster-api-provider-tinkerbell/pull/561) |
| stern/stern | [#364](https://github.com/stern/stern/pull/364) |
| mozillazg/ptcpdump | [#385](https://github.com/mozillazg/ptcpdump/pull/385) |
| Azure/secrets-store-csi-driver-provider-azure | [#2017](https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/2017) |
| ahmetb/gen-crd-api-reference-docs | [#141](https://github.com/ahmetb/gen-crd-api-reference-docs/pull/141) |
| undistro/marvin | [#110](https://github.com/undistro/marvin/pull/110) |
| liquidmetal-dev/cluster-api-provider-microvm | [#353](https://github.com/liquidmetal-dev/cluster-api-provider-microvm/pull/353) |
| oam-dev/cluster-gateway | [#168](https://github.com/oam-dev/cluster-gateway/pull/168) |
| coldzerofear/vgpu-manager | [#45](https://github.com/coldzerofear/vgpu-manager/pull/45) |
| netgroup-polito/CrownLabs | [#1101](https://github.com/netgroup-polito/CrownLabs/pull/1101) |
| GoogleCloudPlatform/k8s-config-connector | [#7662](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/7662) |
| zarf-dev/zarf | [#4846](https://github.com/zarf-dev/zarf/pull/4846) |
| openconfig/kne | [#695](https://github.com/openconfig/kne/pull/695) |
| luxury-yacht/app | [#164](https://github.com/luxury-yacht/app/pull/164) |
| dapr/dapr | [#9712](https://github.com/dapr/dapr/pull/9712) |
| kubernetes/kops | [#18231](https://github.com/kubernetes/kops/pull/18231) |
| kubeedge/kubeedge | [#6748](https://github.com/kubeedge/kubeedge/pull/6748) |
| openkruise/kruise | [#2421](https://github.com/openkruise/kruise/pull/2421) |
| FairwindsOps/goldilocks | [#850](https://github.com/FairwindsOps/goldilocks/pull/850) |
| FairwindsOps/nova | [#454](https://github.com/FairwindsOps/nova/pull/454) |
| kubernetes/dns | [#763](https://github.com/kubernetes/dns/pull/763) |
| kubernetes-sigs/alibaba-cloud-csi-driver | [#1677](https://github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pull/1677) |
| kubesphere/kubekey | [#3086](https://github.com/kubesphere/kubekey/pull/3086) |
| GoogleCloudPlatform/kubectl-ai | [#664](https://github.com/GoogleCloudPlatform/kubectl-ai/pull/664) |
| danielqsj/kafka_exporter | [#519](https://github.com/danielqsj/kafka_exporter/pull/519) |
| kube-logging/logging-operator | [#2230](https://github.com/kube-logging/logging-operator/pull/2230) |

All PRs follow the same pattern: opt into the new `legacy_stderr_threshold_behavior=false` so that `-stderrthreshold` is honored even when `-logtostderr=true` (the default). This sets `stderrthreshold=INFO` as the new default (preserving current behavior) while allowing users to override it with `-stderrthreshold=WARNING` or `-stderrthreshold=ERROR` on the command line.

BenTheElder commented:
@pierluigilenoci I suggest only referring back to this in the PR bodies instead of the commits, it's really hard to follow the history of this thread.
